README


This archive contains the distribution Sub-Chain,
version 0.010012:

  Chain subs together and call in succession

This software is copyright (c) 2011 by Randy Stauner.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision History for Sub-Chain
==============================

0.010012
--------

0270107 fix pod - non-existent push method substituted for existing append method

25bb5e7 Trade 'simple' for 'easy'

74d269d Add TODO Section (but just link to Sub::Chain/TODO)

e2e5a33 Remove 'name_sub' alias

b63aaa7 Add 'tests' and 'documentation' to the TODO list

7338c58 Move Test::Warn suggested test prereq to correct dist.ini

7d17982 Reorganize subs for consistency

0530074 Format the call() input => output documentation in a list for clarity

6d3db6c Add example usage code to groups() documentation

02f4812 Spell out 'or' instead of using a slash

a065a0e Fix module name typo in Pod example

2d0f463 Fix grammatical Pod errors

v0.10 2011-01-04
----------------

0609e32 Move Sub::Chain::Group to a separate distribution

669b30f Add stopwords for pod spelling test

1e5e073 Define missing subs to make test pass again

a97f8f5 Rename $stack to $chain for clarity in test script

b7ea5cb Update transform.t to use new module API

4f3d253 Define new_sub_chain() for extracting 'chain_class' and 'chain_args' usage

f7d703a Explicitly return nothing from dequeue

67c0cc4 Remove grammatical typo from carp message

31d9864 Modify group.t to match module changes

958fe8b Change 'cluck' to 'carp'

5d095a1 Require the specified chain_class before attempting to instantiate it

488c57e Copy chain_class and chain_args from %opts to $self in constructor

c0437ba Make Sub::Chain::Group include itself in Sub::Chain::CARP_NOT

5e53dfe Substitute 'stack' for 'chain' in identifiers in test script

99f7087 Change DTN::Stackable test to Sub::Chain::Group test

493e018 Remove old DTN logic from old stackable test

bfbd005 Change lexical variable name to be more appropriate

c18fffc Remove unnecessary on_undef enum processing

679bba3 Instantiate with the 'chain_class' parameter

61df446 Accept the $sub parameter however they send it

5c39186 Change Pod verbage to sound more like the new module

cc85f78 Update git ignore file to reflect new dist name

dfb9120 Begin converting DTN::Stackable to Sub::Chain::Group

0c450dc Remove unnecessary 'push' alias

62efdca Remove on_undef enum which now exists in Sub::Chain

1860558 Remove remains of DTN now moved to more appropriate modules

d998783 Add Sub::Chain::Named subclass to support appending subs by name

a8fa584 Warn on call() if no subs were added to the chain

23cc0db Rename chain.t to examples.t to be clear that it tests the examples

f8e2f01 Replace small amounts of functionality with new Sub::Chain module

d609583 Add Rationale and See Also Pod

91e355b Add Description Pod and update README and dist.ini to reflect new main module

49b5d4c Add Options section to Pod to describe the enums

9c13323 Add 'result' enum for replacing or discarding the result of each sub

5099557 Copy enum value from $from or $self or $enum (whichever is found first)

835bfda Start testing new Sub::Chain by testing the examples from the POD

v0.9 2011-01-03
---------------

e5222b0 Add initial Sub::Chain module for simple chaining of subs

825c7e4 Reflect return value change of Set::DynamicGroups::items from arrayref to list

09b4765 Reflect Set::DynamicGroups API change from 'append' to 'add'

689cf70 Add stopwords for TODO and Finalizers

6b0f146 Remove module and tests for 'Common' functions that does not belong here

708028e Remove TODO item about transformer options (like on_undef)

6228d0f Test sending undef to the Common functions

98c1d67 Add some tests for 'undef'

d1b06ed Alter tests to match new options hash and array stack interface

c324b9e Drop Test::Exception and test life/death manually due to weird external bug

e6621a4 Alter test to use new %Enums hash instead of the single scalar

1b93396 Alter the test to expect a code ref rather than a Data::Transform::Map instance

f9624f2 Convert hashes of tests to arrays to allow 'undef' as an input value

v0.8 2010-12-27
---------------

597955c Trade Data::Transform[::Stackable] for a simple array of subs/args and then call them sequentially

33545ac Simplify the options to push() by expecting a hash[ref] and sending it through _normalize_spec()

a02d692 Add on_undef Enum and put both enums in our %Enums

4c2af6c Make 'transform' return a sub rather than Data::Transform::Map and rename to 'transformer' to clarify

33d8e9e Add 'consider transformer options' to the TODO list

7f39e52 Define Stackable::named() so that you can add named transformations after instantiation

84332c5 Add TODO list to the bottom of the POD regarding the possibility of dropping Data::Transform

9e661d7 Add comment about 'finalizer' idea (transformation always at the end of the stack)

99f370d Provide groups() method to return the Set::DynamicGroups object

0504e17 Add fields() method for specifying all fields

v0.7 2010-12-23
---------------

eb21ea8 Use Set::DynamicGroups to manage groups and fields instead of handling it internally

99e1f49 Don't die if a transformation is specified for a group that never gets any fields

be3c6d4 Make Common::exchange return the value if no fallback was supplied

3830df1 Test warning and death (if suggested modules are available) for the warn_no_field option

v0.6 2010-12-18
---------------

c4c5d5a Add option for when to warn when a requested field/stack is missing

71d9ad3 Define a helper method on the object instead of generating a dynamic sub for each call to transform()

191acb2 Clarify POD for transform()

c2e3695 Document reprocess_queue() to pass Pod::Coverage test

cbb4b06 Remove made-up word to pass spelling test

522484a Allow Common->_all() to take a list of names (if you don't want /all/)

ceecac6 Fix incorrect Carp imports

b4137a4 Wrap the CORE [lu]c(first)? methods for convenience

514bcc4 Improve usefulness of 'match' function by returning $1 if a group was captured

0e79da0 White box test the queue, the stacks built from it, and the reprocessing of it

v0.5 2010-12-15
---------------


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  'build_requires' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'Module::Build' => '0.3601',
    'Test::More' => '0.96'
  },
  'configure_requires' => {
    'Module::Build' => '0.3601'
  },
  'dist_abstract' => 'Chain subs together and call in succession',
  'dist_author' => [
    'Randy Stauner <rwstauner@cpan.org>'
  ],
  'dist_name' => 'Sub-Chain',
  'dist_version' => '0.010012',
  'license' => 'perl',
  'module_name' => 'Sub::Chain',
  'recommends' => {},
  'recursive_test_files' => 1,
  'requires' => {
    'Carp' => '0',
    'Object::Enum' => '0.072',
    'overload' => '0',
    'perl' => '5.006'
  },
  'script_files' => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Sub/Chain.pm
package Sub::Chain;
BEGIN {
  $Sub::Chain::VERSION = '0.010012';
}
BEGIN {
  $Sub::Chain::AUTHORITY = 'cpan:RWSTAUNER';
}
# ABSTRACT: Chain subs together and call in succession


use strict;
use warnings;
use Carp qw(croak carp);

# enable object to be called like a coderef
use overload
	'&{}' => \&coderef,
	fallback => 1;

use Object::Enum 0.072 ();

our %Enums = (
	result => Object::Enum->new({unset => 0, default => 'replace',
		values => [qw(replace discard)]}),
	on_undef => Object::Enum->new({unset => 0, default => 'proceed',
		values => [qw(skip blank proceed)]}),
);


sub new {
	my $class = shift;
	my %opts = ref $_[0] ? %{$_[0]} : @_;

	my $self = {
		chain => []
	};
	bless $self, $class;

	$self->_copy_enums(\%opts);

	return $self;
}


sub append {
	my ($self, $sub, $args, $opts) = @_;

	# TODO: normalize_spec (better than this):
	$args ||= [];
	$opts ||= {};
	$self->_copy_enums($opts, $opts);

	CORE::push(@{ $self->{chain} }, [$sub, $args, $opts]);
	# allow calls to be chained
	return $self;
}


sub call {
	my ($self, @args) = @_;
	# cache function call
	my $wantarray = wantarray;

	my @chain = @{ $self->{chain} };
	carp("No subs appended to the chain")
		unless @chain;

	foreach my $tr ( @chain ){
		my ($sub, $extra, $opts) = @$tr;
		my @all = (@args, @$extra);
		my @result;

		# TODO: instead of duplicating enum objects do %opts = (%$self, %$opts)
		if( @args && $opts->{on_undef} && !defined($args[0]) ){
			next if $opts->{on_undef}->is_skip;
			$args[0] = ''
				if $opts->{on_undef}->is_blank;
		}

		# call sub with same context as this
		if( !defined $wantarray ){
			$sub->(@all);
		}
		elsif( $wantarray ){
			@result    = $sub->(@all);
		}
		else {
			$result[0] = $sub->(@all);
		}
		@args = @result
			if $opts->{result}->is_replace;
	}

	# if 'result' isn't 'replace' what would be a good return value?
	# would they expect one?

	# return value appropriate for context
	if( !defined $wantarray ){
		return;
	}
	elsif( $wantarray ){
		return @args;
	}
	else {
		return $args[0];
	}
}


sub coderef {
	my ($self) = @_;
	return sub { $self->call(@_); }
}

sub _copy_enums {
	my ($self, $from, $to) = @_;
	$to ||= $self;
	while( my ($name, $enum) = each %Enums ){
		$to->{$name} = ($self->{$name} || $enum)->clone(
			# use the string passed in
			exists $from->{$name} ? $from->{$name} :
				# clone from the default value saved on the instance
				$self->{$name} ? $self->{$name}->value : ()
		);
	};
}

1;


# TODO: link to questions on perlmonks and stackoverflow?


__END__
=pod

=for :stopwords Randy Stauner runtime distros TODO CPAN AnnoCPAN RT CPANTS Kwalitee diff
IRC

=head1 NAME

Sub::Chain - Chain subs together and call in succession

=head1 VERSION

version 0.010012

=head1 SYNOPSIS

	my $chain = Sub::Chain->new();

	$chain->append(\&wash, ['cold']);
	$chain->append(\&dry,  [{tumble => 'low'}]);
	$chain->append(\&fold);

	my @clean_laundry = $chain->call(@clothes);

	# if only it were that easy

=head1 DESCRIPTION

This module aims to provide a simple interface for
chaining multiple subs (coderefs) together
and executing them one after the other in a single call.

It was specifically designed to be built dynamically
from a list of specifications provided at runtime
to filter data through the specified list of functions.

Also see L<Sub::Chain::Named>
which appends subs to the chain by name rather than coderef.

=head1 METHODS

=head2 new

	my $chain = Sub::Chain->new();
	my $chain = Sub::Chain->new( option => $value );
	my $chain = Sub::Chain->new({option => $value});

Constructor.
Takes a hash or hashref of arguments.

Accepts values as described in L<OPTIONS|/OPTIONS>
that will be used as defaults
for any sub that doesn't override them.

=head2 append

	$chain->append(\&sub, \@args, \%opts);

Append a sub to the chain.
The C<\@args> arrayref will be flattened and passed to the C<\&sub>
after any arguments to L</call>.

	sub sum { my $s = 0; $s += $_ for @_; $s; }

	$chain->append(\&sum, [3, 4]);

	$chain->call(1, 2);
	# returns 10
	# equivalent to: sum(1, 2, 3, 4)

If you don't want to send any additional arguments to the sub
an empty arrayref (C<[]>) can be used.

This method returns the object so that it can be chained for simplicity:

	$chain->append(\&sub, \@args)->append(\&sub2)->append(\&sub3, [], \%opts);

The C<\%opts> hashref can be any of the options described in L</OPTIONS>
to override the defaults on the object for this particular sub.

=head2 call

	$chain->call(@args);

Calls each method in the chain
with the supplied (and any predetermined) arguments
according to any predefined options.

=head2 coderef

	my $sub = $chain->coderef;
	$sub->(@args);

Wrap C<< $self->call >> in a closure.
This is used to overload the function dereference operator
so you can pretend the instance is a coderef: C<< $chain->(@args) >>

=head1 OPTIONS

These options can define how a sub should be handled.
Specified in the options hashref for L</append>
they apply to that particular sub.
Specified in the constructor they can set the default
for how to handle any sub that doesn't override the option.

=over 4

=item *

C<result>

What to do with the result;
Valid values are:

=over 4

=item *

C<replace> - replace the argument list with the return value of each sub

=item *

C<discard> - discard the return value of each sub

=back

The arguments to L</call> are passed to each sub in the chain.
When C<replace> is specified the return value of one sub
is the argument list to the next.
This is useful, for instance, when chaining a number of
data cleaning or transformation functions together:

	sub add_uc { $_[0] . ' ' . uc $_[0]  }
	sub repeat { $_[0] x $_[1] }

	$chain->append(\&add_uc)->append(\&repeat, [2]);
	$chain->call('hi');

	# returns 'hi Hihi HI', similar to:

	my $s = 'hi';
	$s = add_uc($s);
	$s = repeat($s, 2);

When C<discard> is specified, the same arguments are sent to each sub.
This is useful when chaining subs that are called for their side effects
and you aren't interested in the return values.

	# assume database handle has RaiseError set
	$chain
		->append(\&log)
		->append(\&save_to_database)
		->append(\&email_to_user);

	# call in void context because we don't care about the return value
	$chain->call($object);

The default is C<replace> since that (arguably) makes this module more useful.

=item *

C<on_undef>

What to do when a value is undefined;
Valid values are:

=over 4

=item *

C<proceed> - proceed as normal (as if it was defined)

=item *

C<skip> - skip (don't call) the sub

=item *

C<blank> - initialize the value to a blank string

=back

The default is C<proceed>.

=back

=head1 RATIONALE

This module started out as C<Data::Transform::Named>,
a named wrapper (like C<Sub::Chain::Named>) around
L<Data::Transform> (and specifically L<Data::Transform::Map>).

As the module was nearly finished I realized I was using very little
of L<Data::Transform> (and its documentation suggested that
I probably wouldn't want to use the only part that I I<was> using).
I also found that the output was not always what I expected.
I decided that it seemed reasonable according to the likely purpose
of L<Data::Transform>, and this module simply needed to be different.

So I attempted to think more abstractly
and realized that the essence of the module was not tied to
data transformation, but merely the succession of simple subroutine calls.

I then found and considered L<Sub::Pipeline>
but needed to be able to use the same
named subroutine with different arguments in a single chain,
so it seemed easier to me to stick with the code I had written
and just rename it and abstract it a bit further.

I also looked into L<Rule::Engine> which was beginning development
at the time I was searching.
However, like L<Data::Transform>, it seemed more complex than what I needed.
When I saw that L<Rule::Engine> was using [the very excellent] L<Moose>
I decided to pass since I was doing work on a number of very old machines
with old distros and old perls and constrained resources.
Again, it just seemed to be much more than what I was looking for.

=head1 TODO

=over 4

=item *

Write a lot more tests

=item *

Improve documentation

=back

=head1 SEE ALSO

=over 4

=item *

L<Sub::Chain::Named>

=item *

L<Sub::Pipeline>

=item *

L<Data::Transform>

=item *

L<Rule::Engine>

=back

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

  perldoc Sub::Chain

=head2 Websites

The following websites have more information about this module, and may be of help to you. As always,
in addition to those websites please use your favorite search engine to discover more resources.

=over 4

=item *

Search CPAN

L<http://search.cpan.org/dist/Sub-Chain>

=item *

RT: CPAN's Bug Tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Sub-Chain>

=item *

AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Sub-Chain>

=item *

CPAN Ratings

L<http://cpanratings.perl.org/d/Sub-Chain>

=item *

CPAN Forum

L<http://cpanforum.com/dist/Sub-Chain>

=item *

CPANTS Kwalitee

L<http://cpants.perl.org/dist/overview/Sub-Chain>

=item *

CPAN Testers Results

L<http://cpantesters.org/distro/S/Sub-Chain.html>

=item *

CPAN Testers Matrix

L<http://matrix.cpantesters.org/?dist=Sub-Chain>

=back

=head2 Bugs / Feature Requests

Please report any bugs or feature requests by email to C<bug-sub-chain at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Sub-Chain>.  I will be
notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head2 Source Code


L<http://github.com/magnificent-tears/Sub-Chain/tree>

  git clone git://github.com/magnificent-tears/Sub-Chain.git

=head1 AUTHOR

Randy Stauner <rwstauner@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Randy Stauner.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Sub/Chain/Named.pm
package Sub::Chain::Named;
BEGIN {
  $Sub::Chain::Named::VERSION = '0.010012';
}
BEGIN {
  $Sub::Chain::Named::AUTHORITY = 'cpan:RWSTAUNER';
}
# ABSTRACT: subclass of Sub::Chain with named subs


use strict;
use warnings;
use Carp qw(croak);
use Sub::Chain;
our @ISA = qw(Sub::Chain);


sub new {
	my $class = shift;
	my %opts = ref $_[0] ? %{$_[0]} : @_;
	my $subs = delete $opts{subs};

	my $self = $class->SUPER::new(%opts);
	$self->{named} = $subs || {};

	return $self;
}


sub append {
	my ($self, $name, @append) = @_;
	my $sub = $self->{named}{$name}
		or croak("No sub defined for name: $name");
	$self->SUPER::append($sub, @append);
}


sub name_subs {
	my ($self) = shift;
	my %subs = ref $_[0] ? %{$_[0]} : @_;

	# TODO: warn if already exists?
	@{ $self->{named} }{keys %subs} = values %subs;

	# chainable
	return $self;
}

1;


__END__
=pod

=for :stopwords Randy Stauner

=head1 NAME

Sub::Chain::Named - subclass of Sub::Chain with named subs

=head1 VERSION

version 0.010012

=head1 SYNOPSIS

	my $chain = Sub::Chain::Named->new(subs => {name1 => \&sub1});
	$chain->name_subs(name2 => \&sub2, name3 => \&sub3);
	...
	$chain->append($name, \@args, \%opts);

=head1 DESCRIPTION

This is a subclass of L<Sub::Chain>.
It stores a list of named subs
and then accepts the name as an argument to L</append>
(instead of the coderef).

This can simplify things if, for example,
you are taking the list of subs dynamically from file input.

=head1 METHODS

=head2 new

	my $chain = Sub::Chain::Named->new(
		subs => {
			action => sub {},
		}
	);

Instantiate a L<Sub::Chain> instance
with a collection of named subs.

A hashref of C<< name => \&sub >> pairs can be passed
as the C<subs> option.

See L<Sub::Chain/new> for more information.

=head2 append

	$named->append($sub_name);
	$named->append($sub_name, \@args, \%opts);

Just like L<Sub::Chain/append>
except that C<$sub_name> is a string
which is converted to the corresponding sub
and then passed to L<Sub::Chain/append>.

=head2 name_subs

	$named->name_subs(goober => \&peant_butter);

Add named subs to the collection.
Takes a hash (or hashref),
or just a single name and a value (a small hash).

=head1 AUTHOR

Randy Stauner <rwstauner@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Randy Stauner.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

BEGIN { require 5.006; }

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'Chain subs together and call in succession',
  'AUTHOR' => 'Randy Stauner <rwstauner@cpan.org>',
  'BUILD_REQUIRES' => {
    'File::Find' => '0',
    'File::Temp' => '0',
    'ExtUtils::MakeMaker' => '6.31',
    'Test::More' => '0.96'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'Sub-Chain',
  'EXE_FILES' => [],
  'LICENSE' => 'perl',
  'NAME' => 'Sub::Chain',
  'PREREQ_PM' => {
    'Carp' => '0',
    'Object::Enum' => '0.072',
    'overload' => '0'
  },
  'VERSION' => '0.010012',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



